### ISSUE TITLE
switchOnFirst multithread issue?

### ISSUE THREAD

**Issue by koldat** [reactor/reactor-core#2715]
We have upgraded Spring Boot to 2.4.6 (form 2.3.x). It is including reactor 3.4.6. From that point we see time to time some strange NPE that I do not see how it can happen. I see that there was a recent change in switchOnFirst related to threading: #2671 #1496 We also use it with rsocket.

## Expected Behavior
switchOnFirst should not pass null to downstream and handle the switch properly with multiple threads (publishOn, subscribeOn).

## Actual Behavior
```java
    public <R, M> Flux<R> buildRequestChannel(Publisher<? extends M> request, 
        Function<M, ByteBuffer> encoder, Function<Payload, R> decoder) {
        return 
            rSocket.flatMapMany(socket -> socket.requestChannel(
                Flux.from(request)
                .publishOn(scheduler)
                .map(encoder) <- Creates ByteBuffer (impossible to be null as it is ByteBuffer.wrap())
                .switchOnFirst((signal, inner) -> {
                    if (signal.hasValue()) {
                        Payload firstPayload = DefaultPayload.create(signal.get(), encodeMetadata());
                        return Mono.just(firstPayload)
                            .concatWith(inner.skip(1).map(DefaultPayload::create));
                    }
                    return inner.map(DefaultPayload::create);
                })
                .transform(f -> transformSend.apply(f))) <-- HERE
            .publishOn(scheduler)
            .transform(f -> transformReceive.apply(f))
            .map(decoder))
            .onErrorMap(applicationExceptionMapper())
            .contextWrite(ctx -> ctx.put(ORIG_URL, target));
    }
```

HERE line is our monitoring transformer where e simply count bytes of DefaultPayload, but data is null. We call there that getData(). The only possibility is to get null to DefaultPayload.create in the code you see (upstream cannot generate null.) 

FluxSwitchOnFirst is playing a lot with variables around "firstValue" which are not volatile so maybe it 

```
java.lang.NullPointerException: Cannot invoke "java.nio.ByteBuffer.duplicate()" because "this.data" is null
    at io.rsocket.util.DefaultPayload.getData(DefaultPayload.java:138)
    at com.ca.apm.common.rpc.RSocketServices.lambda$19(RSocketServices.java:516)
    at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:184)
    at reactor.core.publisher.FluxSwitchOnFirst$SwitchOnFirstControlSubscriber.onNext(FluxSwitchOnFirst.java:853)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:177)
    at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:120)
    at reactor.core.publisher.FluxSkip$SkipSubscriber.onNext(FluxSkip.java:86)
    at reactor.core.publisher.FluxSwitchOnFirst$SwitchOnFirstMain.tryDirectSend(FluxSwitchOnFirst.java:728)
<-- HERE firstValue was resolved to null ???
    at reactor.core.publisher.FluxSwitchOnFirst$AbstractSwitchOnFirstMain.sendFirst(FluxSwitchOnFirst.java:686)
    at reactor.core.publisher.FluxSwitchOnFirst$AbstractSwitchOnFirstMain.request(FluxSwitchOnFirst.java:630)
    at reactor.core.publisher.FluxSkip$SkipSubscriber.onSubscribe(FluxSkip.java:78)
    at reactor.core.publisher.FluxSwitchOnFirst$AbstractSwitchOnFirstMain.subscribe(FluxSwitchOnFirst.java:672)
    at reactor.core.publisher.Flux.subscribe(Flux.java:8235)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:208)
    at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:80)
    at reactor.core.publisher.Flux.subscribe(Flux.java:8235)
<-- HERE firstValue cannot be null
    at reactor.core.publisher.FluxSwitchOnFirst$AbstractSwitchOnFirstMain.onNext(FluxSwitchOnFirst.java:483)
    at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:127)
    at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runAsync(FluxPublishOn.java:440)
    at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:527)
    at reactor.core.scheduler.ImmediateScheduler$ImmediateSchedulerWorker.schedule(ImmediateScheduler.java:83)
    at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.trySchedule(FluxPublishOn.java:312)
    at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.onNext(FluxPublishOn.java:237)
    at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:127)
    at reactor.core.publisher.FluxGroupBy$UnicastGroupedFlux.drainRegular(FluxGroupBy.java:567)
    at reactor.core.publisher.FluxGroupBy$UnicastGroupedFlux.drain(FluxGroupBy.java:652)
    at reactor.core.publisher.FluxGroupBy$UnicastGroupedFlux.subscribe(FluxGroupBy.java:718)
    at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:62)
    at reactor.core.publisher.FluxDeferContextual.subscribe(FluxDeferContextual.java:57)
    at reactor.core.publisher.Flux.subscribe(Flux.java:8235)
    at io.rsocket.core.RequestChannelRequesterFlux.subscribe(RequestChannelRequesterFlux.java:115)
```

## Steps to Reproduce
I am not able to reproduce it in a test. Only our application code.

## Possible Solution
No idea yet. Maybe volatile variables for: isInboundRequestedOnce, isFirstOnNextReceivedOnce, firstValue.

## Your Environment
Spring Boot to 2.4.6 (form 2.3.x)
Reactor Core 3.4.6
RSocket 1.1.0

**Comment by koldat** [comment]
@OlegDokuka can you please take a look as you were involved in recent changes around this operator?

**Comment by OlegDokuka** [comment]
@koldat do you know if the actual upstream has value in that scenario or not?

**Comment by OlegDokuka** [comment]
@koldat do you have anything special in your `transformSend.apply(f)`?

**Comment by OlegDokuka** [comment]
In general, it seems there is `request` race somewhere which causes this problem. `request` race is prohibited by spec and I will investigate the root cause of this race. Thank you for reporting that!

so far, from what I see in your code - you don't need `swiftOnFirst` operator at all. Feel free to simplify your solution to something like that:

```java
public <R, M> Flux<R> buildRequestChannel(Publisher<? extends M> request, 
        Function<M, ByteBuffer> encoder, Function<Payload, R> decoder) {
        return 
            rSocket.flatMapMany(socket -> socket.requestChannel(
                Flux.from(request)
                .publishOn(scheduler)
                .map(encoder) <- Creates ByteBuffer (impossible to be null as it is ByteBuffer.wrap())
                .map(new Function<ByteBuffer, Payload>() {

                    boolean once;

                    public Payload apply(ByteBuffer data) {
                        if (once) {
                           return DefaultPayload.create(data);
                        }

						once = true;
	
                        return DefaultPayload.create(signal.get(), encodeMetadata());
                    }
                })
                .transform(f -> transformSend.apply(f))) <-- HERE
            .publishOn(scheduler)
            .transform(f -> transformReceive.apply(f))
            .map(decoder))
            .onErrorMap(applicationExceptionMapper())
            .contextWrite(ctx -> ctx.put(ORIG_URL, target));
    }
```

**Comment by OlegDokuka** [comment]
@simonbasle from the issue it seems that there is `subscribeOn` after `switchOnFirst`. In that scenario, `FluxSkip` requests n (remaining) elements from `onSubscribe` which can race with subscribeOn#request which causes that issue. Likely fix has to be delivered somewhere over there

**Comment by OlegDokuka** [comment]
can be related #2165

**Comment by koldat** [comment]
@OlegDokuka yes, I have created workarounds on many places to replace switchOnFirst. This example is simple to change, but some are tricky. Especially when you need to emit more values instead of that first one (I can easily use any publisher, concat, etc.)  following by the rest.

Anyway I think your workaround does not work well as it uses "once" from the time we are building the pipeline so it cannot be reused. I am using transform + defer combo as defer generates that boolean per subscription.

Regarding transformSend.apply(f). There is nothing special (just our profiling code):
```java
       upstream -> Flux.deferContextual(ctx -> {
            if (!ctx.hasKey(ProfilingContext.class)) {
                return Flux.from(upstream);
            }

            IOCounter counters = getCounters(ctx);
            return Flux.from(upstream)
                .doOnNext(p -> counters.addToOutput(p.getData().remaining()))
                .doFinally(s -> {
                    counters.flush();
                });
        })
```

**Comment by koldat** [comment]
And yes we use Executor Schedulers :(

The reason was an issue in boundedElastic (I did not create defect yet). When Bounded elastic worker is canceled (generates InterruptedException) is assumes thread is done. But that thread can still be running. Worker is reported as free to use and is used for new subscription. Such a subscription is then doing nothing until original thread finally stops which was in our case long time. Maybe it is not the case anymore, but I haven't tested.

**Comment by OlegDokuka** [comment]
> Anyway I think your workaround does not work well as it uses "once" from the time we are building the pipeline so it cannot be reused. I am using transform + defer combo as defer generates that boolean per subscription.

Yes. Defer should fix that problem.

> And yes we use Executor Schedulers :(

That is (probably) the root cause. The operator follows the reactive streams specification which stays that both onNext onError onComplete must be executed serially to each other. And the same rule applicable to subscription request | cancel method so they must be executed sequentially. Thus the impl just simply expects the correctness of that which could be broken in case of subscribeOn + Scheduler.fromExecutor(...)

I will be working toward the fix for #2165. 

As an alternative solution, I can suggest writing your own Scheduler interface impl which ensures sequential task execution.

**Comment by koldat** [comment]
@OlegDokuka Can you check #2716? It runs as FIFO using this idea: https://stackoverflow.com/questions/29889885/java-divide-incoming-work-uniformly-via-hashing-in-multithreaded-evnironments/
